File: cli/src/core/references/sources/HtmlEntities.test.ts
import { describe, expect, it } from "vitest";
import { decodeHtmlEntities } from "./HtmlEntities.js";

describe("decodeHtmlEntities", () => {
	it("decodes the five known named entities", () => {
		expect(decodeHtmlEntities("&amp; &lt; &gt; &quot; &apos;")).toBe("& < > \" '");
	});

	it("decodes decimal numeric entities", () => {
		// &#960; → π (U+03C0)
		expect(decodeHtmlEntities("Hello &#960; world")).toBe("Hello π world");
	});

	it("decodes hex numeric entities (lowercase x)", () => {
		// &#x2026; → … (HORIZONTAL ELLIPSIS).
		// HTML spec allows uppercase `&#X…;` too, but real GitHub MCP output and
		// the regex used here are both lowercase-only — keeping the surface
		// narrow avoids tolerating malformed entities.
		expect(decodeHtmlEntities("ellipsis: &#x2026;")).toBe("ellipsis: …");
	});

	it("passes through unknown named entities unchanged (defensive — no silent corruption)", () => {
		expect(decodeHtmlEntities("&foo; &nbsp; &copy;")).toBe("&foo; &nbsp; &copy;");
	});

	it("passes through out-of-range hex code points unchanged (no throw)", () => {
		// &#x110000; is one past the Unicode max (U+10FFFF). String.fromCodePoint would
		// throw RangeError; the decoder's range guard returns the original string instead.
		expect(decodeHtmlEntities("oob: &#x110000;")).toBe("oob: &#x110000;");
	});

	it("passes through out-of-range decimal code points unchanged (no throw)", () => {
		// 1114112 = 0x110000 — same out-of-range guard, decimal path.
		expect(decodeHtmlEntities("oob: &#1114112;")).toBe("oob: &#1114112;");
	});

	it("leaves bare strings without entities alone", () => {
		expect(decodeHtmlEntities("plain text with no entities")).toBe("plain text with no entities");
	});

	it("handles consecutive entities and entity-rich strings", () => {
		expect(decodeHtmlEntities("&amp;&amp;&amp;")).toBe("&&&");
		expect(decodeHtmlEntities("&lt;tag attr=&quot;v&quot;&gt;")).toBe('<tag attr="v">');
	});

	it("decodes ASCII control range (e.g. newline) via decimal", () => {
		// &#10; → newline; in-range, normal path.
		expect(decodeHtmlEntities("a&#10;b")).toBe("a\nb");
	});

	it("decodes ASCII control range via hex", () => {
		// &#x0A; → newline
		expect(decodeHtmlEntities("a&#x0A;b")).toBe("a\nb");
	});

	it("passes through lone UTF-16 surrogate code points unchanged (hex)", () => {
		// U+D800 is a high surrogate. String.fromCodePoint does NOT throw on it —
		// it would emit a lone surrogate and silently corrupt the text — so the
		// guard must leave the entity verbatim.
		expect(decodeHtmlEntities("x&#xD800;y")).toBe("x&#xD800;y");
		expect(decodeHtmlEntities("x&#xDFFF;y")).toBe("x&#xDFFF;y");
	});

	it("passes through a lone surrogate code point unchanged (decimal)", () => {
		// 55296 = 0xD800
		expect(decodeHtmlEntities("x&#55296;y")).toBe("x&#55296;y");
	});

	it("still decodes the code points just outside the surrogate block", () => {
		// U+D7FF (just below) and U+E000 (just above) are valid scalar values.
		expect(decodeHtmlEntities("&#xD7FF;")).toBe(String.fromCodePoint(0xd7ff));
		expect(decodeHtmlEntities("&#xE000;")).toBe(String.fromCodePoint(0xe000));
	});
});

File: intellij/src/test/kotlin/ai/jolli/jollimemory/core/references/HtmlEntitiesTest.kt
package ai.jolli.jollimemory.core.references

import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class HtmlEntitiesTest {

	@Nested
	inner class NamedEntities {
		@Test
		fun `decodes amp lt gt quot apos`() {
			HtmlEntities.decode("&amp; &lt; &gt; &quot; &apos;") shouldBe "& < > \" '"
		}

		@Test
		fun `passes through unknown named entities`() {
			HtmlEntities.decode("&nbsp; &mdash;") shouldBe "&nbsp; &mdash;"
		}
	}

	@Nested
	inner class NumericEntities {
		@Test
		fun `decodes decimal entities`() {
			HtmlEntities.decode("&#960;") shouldBe "π"
		}

		@Test
		fun `decodes hex entities`() {
			HtmlEntities.decode("&#x2026;") shouldBe "…"
		}

		@Test
		fun `decodes newline`() {
			HtmlEntities.decode("&#10;") shouldBe "\n"
			HtmlEntities.decode("&#x0A;") shouldBe "\n"
		}

		@Test
		fun `rejects out-of-range code points`() {
			HtmlEntities.decode("&#x110000;") shouldBe "&#x110000;"
		}

		@Test
		fun `rejects surrogate code points`() {
			HtmlEntities.decode("&#xD800;") shouldBe "&#xD800;"
			HtmlEntities.decode("&#xDFFF;") shouldBe "&#xDFFF;"
		}
	}

	@Nested
	inner class Mixed {
		@Test
		fun `handles consecutive entities`() {
			HtmlEntities.decode("&lt;&gt;") shouldBe "<>"
		}

		@Test
		fun `no-op on plain text`() {
			HtmlEntities.decode("hello world") shouldBe "hello world"
		}

		@Test
		fun `empty string`() {
			HtmlEntities.decode("") shouldBe ""
		}
	}
}
